music_assistant/server/controllers/player_queues.py

```diff
@@ -1050,6 +1081,15 @@ def on_player_update(
                 and (queue.items - queue.current_index) < 5
             ):
                 self.mass.create_task(self._fill_radio_tracks(queue_id))
+            elif (
+                # We have received the last item in the queue and Don't stop the music is enabled
+                queue.dont_stop_the_music_enabled
+                and queue.current_index
+                and (queue.items - queue.current_index) <= 1
+            ):
+                # Enable radio mode on the originally requested MediaItem in play_media
+                queue.radio_source = queue.played_media_item
```

Comment:
Isn't this a super complicated way of doing this ?
We have all the already played tracks available in the queue_items

Reply:
Well we do not want the played tracks as a source for our radio. We only want to use the _originally enqueued media items_ as a source for the radio, because they represent the closest of what the user intended to play. 

The logic here checks if we are playing the last item in the queue and, if DSTM is enabled, uses those _originally enqueued media items_ as a radio source to dynamically fill the queue.

Reply:
Sure but radio mode is always based on the tracks right ?
So radio mode for playlist --> will be based on the tracks
So radio mode for album--> will be based on the tracks

Reply:
> Sure but radio mode is always based on the tracks right ? So radio mode for playlist --> will be based on the tracks So radio mode for album--> will be based on the tracks

No, radio mode is based on `MediaItems`. So the `radio_source` can be a list of let's say `[Album, Track, Playlist]`. The radio _logic_ translates this into a list of base tracks (random 5 tracks per `MediaItem`), in this case 5 album tracks + the enqueued track + 5 random playlist tracks. Then it will create the dynamic playlist based on similar tracks for all of those base tracks.

This ensures that the originally enqueued items are remembered so we can always go back to them for generating a fresh list of dynamic tracks.